### PR TITLE
Rooms bug fixes and tests

### PR DESCRIPTION
**Pull request by alexrosen45**

### CODE REVIEW

File: src/use_case/rooms/LoadRoomsInteractor.java [hunk 1]
```diff
@@ -27,9 +27,6 @@ public void loadRooms(LoadRoomsInputData roomsInputData) {
       List<Room> rooms = new ArrayList<>();
       for (String roomId : availableRoomIds) {
         Room room = roomsDataAccessObject.getRoomFromId(roomsInputData.user, userDao, roomId);
-        System.out.println(room.getUid());
```
Comment (uoftm) [inline src/use_case/rooms/LoadRoomsInteractor.java:30]: I could've sworn this is deleted in main

File: tests/view/RoomsTest.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+    RoomsState testState = buildTestState();
+    testState.setUserToAddEmail(createDummyDisplayUser().getEmail());
+    roomsViewModel.setState(testState);
+
+    JButton addUserButton = roomsView.getAddUserButton();
+    addUserButton.doClick();
+
+    String regex =
+        "Unable to add \\s*([a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,})\\s*to room:";
```
Comment (uoftm) [inline tests/view/RoomsTest.java:140]: why use a complicated regex if you're not going to go [all the way](https://emailregex.com/)

File: tests/view/RoomsTest.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+    RoomsViewModel roomsViewModel = new RoomsViewModel();
+    RoomsView roomsView =
+        RoomsUseCaseFactory.create(
+            new FirebaseRoomsDataAccessObject(client),
+            new FirebaseMessageDataAccessObject(client),
+            new FirebaseUserDataAccessObject(client),
+            roomsViewModel,
+            searchController,
+            null);
```
Comment (uoftm) [inline tests/view/RoomsTest.java:103]: interesting and a bit worrying that this works, but ok

File: tests/view/RoomsTest.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    RoomsState testState = buildTestState();
+    testState.setRoomUid(testState.getAvailableRooms().get(0).getUid());
+    testState.setSendMessage("Test message!");
+    System.out.println(testState.getAvailableRooms().get(0).getMessages());
+    roomsViewModel.setState(testState);
+
+    JButton sendMessageButton = roomsView.getSendMessageButton();
+    sendMessageButton.doClick();
+
+    System.out.println(roomsViewModel.getState().getSuccess());
```
Comment (uoftm) [inline tests/view/RoomsTest.java:114]:
```suggestion
```

File: tests/view/RoomsTest.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+            new FirebaseMessageDataAccessObject(client),
+            new FirebaseUserDataAccessObject(client),
+            roomsViewModel,
+            searchController,
+            null);
+
+    RoomsState testState = buildTestState();
+    testState.setRoomUid(testState.getAvailableRooms().get(0).getUid());
+    testState.setSendMessage("Test message!");
+    System.out.println(testState.getAvailableRooms().get(0).getMessages());
```
Comment (uoftm) [inline tests/view/RoomsTest.java:108]:
```suggestion
```

Ideally tests should only log on failure

File: src/data_access/FirebaseMessageDataAccessObject.java [hunk 6]
```diff
@@ -21,7 +21,11 @@ public void sendMessage(
       Room room, LoginUserDataAccessInterface userDAO, User user, String messageBody) {
     String idToken = userDAO.getAccessToken(user.getEmail(), user.getPassword());
     String messageJSON =
-        new JSONObject().put("contents", messageBody).put("author", user.getEmail()).toString();
+        new JSONObject()
+            .put("contents", messageBody)
+            .put("displayName", user.getEmail())
```
Comment (uoftm) [inline src/data_access/FirebaseMessageDataAccessObject.java:26]: why do we need this change?

File: tests/view/RoomsTest.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+  Room createDummyRoom() {
+    DisplayUser dummyDisplayUser = new DisplayUser("dummyUid", "Dummy User");
+    List<DisplayUser> users = new ArrayList<>();
+    users.add(dummyDisplayUser);
+    List<Message> messages = new ArrayList<>();
+    messages.add(new Message(Instant.now(), "Dummy message", dummyDisplayUser));
+    Room room = new Room("dummyRoomUid", "Dummy Room", users, null);
+    room.setMessages(messages);
+    return room;
+  }
```
Comment (csc207jc) [inline tests/view/RoomsTest.java:177]: There are already functions that do this in `DAOTest.java`.
Maybe call those instead?

File: tests/view/RoomsTest.java (hunk 3 above)
Comment (alexrosen45) [reply tests/view/RoomsTest.java:103]: ?

File: tests/view/RoomsTest.java (hunk 7 above)
Comment (alexrosen45) [reply tests/view/RoomsTest.java:176]: Yes! I was going to restructure the tests file structure to do this properly, but other people are still committing tests so I left it like this for now. Should I add this change now?

File: src/data_access/FirebaseMessageDataAccessObject.java (hunk 6 above)
Comment (alexrosen45) [reply src/data_access/FirebaseMessageDataAccessObject.java:26]: To display the usernames of users in the rooms chat

File: tests/view/RoomsTest.java (hunk 3 above)
Comment (uoftm) [reply tests/view/RoomsTest.java:103]: (Passing the null in instead of a controller)

File: tests/view/RoomsTest.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+    RoomsState testState = buildTestState();
+    testState.setRoomUid(testState.getAvailableRooms().get(0).getUid());
+    testState.setSendMessage("Test message!");
+    roomsViewModel.setState(testState);
+
+    JButton sendMessageButton = roomsView.getSendMessageButton();
+    sendMessageButton.doClick();
+
+    Assert.assertTrue(roomsViewModel.getState().getError() != null);
```
Comment (uoftm) [inline tests/view/RoomsTest.java:113]: I think there should be an assert function for assertNonNull?

File: tests/view/RoomsTest.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+            null,
+            null);
+
+    roomsViewModel.setState(this.buildTestState());
+
+    JButton refresh = roomsView.getRefreshButton();
+    refresh.doClick();
+
+    // User can load blank rooms with request to Firebase
+    Assert.assertTrue(roomsViewModel.getState().getError() == null);
```
Comment (uoftm) [inline tests/view/RoomsTest.java:53]: Similarly, I believe you can do assertNull here

Comment (alexrosen45) [reply tests/view/RoomsTest.java:53]: ok changing this

**Review by uoftm — APPROVED** [review]

**Pull request merged**